### PR TITLE
Fix lexer cache

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
This fixes https://github.com/Microsoft/visualfsharp/issues/2155

It reverts #2090, then fixes the `OutOfRangeException` in `getCachedSourceLineData`.

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
OMG, this code is... let's call it fragile. Please, don't touch it ever in the future.

**Comment by cartermp** [comment]
Perhaps in the future we'll be able to do away with all of these `+1`/`-1` offsets and make FCS 0-based...

**Comment by vasily-kirichenko** [comment]
@dsyme Maybe rename `fromZ` and `toZ` to `fromVS` and `toVS`? And `ZeroBasedLineAnnotation` to `VSLineNumber`?

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko this stuff is in FCS / not specific to VS and I think if it was to be renamed `fromZero` and `toZero` would be the correct names.

**Comment by vasily-kirichenko** [comment]
@smoothdeveloper As there are other zero line based editors, then yes, the functions should keep their names, but we should create aliases in FSharp.Editor project with the proposed names.

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by KevinRansom** [comment]
@vasily-kirichenko 

````
1) Failed : Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.GoToDefinitionServiceTests.VerifyDefinition
Incorrect information returned for fileContents=<<<
type TestType() =
    member this.Member1(par1: int) =
        printf "%d" par1
    member this.Member2(par2: string) =
        printf "%s" par2

[<EntryPoint>]
let main argv =
    let obj = TestType()
    obj.Member1(5)
    obj.Member2("test")>>>, caretMarker=<<<obj.Member1>>>, expected =<<<Some (3, 3, 16, 23)>>>, actual = <<<<null>>>>
at Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.GoToDefinitionServiceTests.VerifyDefinition() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\GoToDefinitionServiceTests.fs:line 108

2) Failed : Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.QuickInfoProviderTests.ShouldShowQuickInfoAtCorrectPositions
  Expected: <Some(val x : int
Full name: Test.x)>
  But was:  null
at Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.QuickInfoProviderTests.ShouldShowQuickInfoAtCorrectPositions() in D:\j\workspace\release_ci_pa---3f142ccc\vsintegration\tests\unittests\QuickInfoProviderTests.fs:line 105

Run Settings
    RuntimeFramework: V4.0
    RunAsX86: True
    WorkDirectory: D:\j\workspace\release_ci_pa---3f142ccc\release\net40\bin
    MaxAgents: 1
    NumberOfTestWorkers: 1
    Verbose: True

Test Run Summary
    Overall result: Failed
   Tests run: 2936, Passed: 2934, Errors: 0, Failures: 2, Inconclusive: 0
     Not run: 99, Invalid: 0, Ignored: 99, Explicit: 0, Skipped: 0
  Start time: 2017-01-05 07:14:39Z
    End time: 2017-01-05 07:31:22Z
    Duration: 1003.057 seconds


````

**Comment by vasily-kirichenko** [comment]
@KevinRansom fixed.

File: vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs [hunk 1]
```diff
@@ -133,72 +128,66 @@ module internal CommonHelpers =
...
+ 
+             let startLine = lines.GetLineFromPosition(textSpan.Start).LineNumber
+             let endLine = lines.GetLineFromPosition(textSpan.End).LineNumber
+             // Go backwards to find the last cached scanned line that is valid
+             let scanStartLine = 
+                 let mutable i = startLine
+                 while i > 0 && (match sourceTextData.[i] with Some data -> not (data.IsValid(lines.[i])) | None -> true)  do
+                     i <- i - 1
+                 i
+             // Rescan the lines if necessary and report the information
```
Comment (KevinRansom) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: scanStartLine made my head explode or at least the while loop condition did.
I don't know whether this is any more readable though:
````
             let scanStartLine =
                 let rec scanStart i =
                     if i > 0 && (Option.isNone (sourceTextData.[i]) || not (sourceTextData.[i].Value.IsValid(lines.[i]))) then scanStart (i - 1)
                     else i
                 scanStart startLine
````

Comment (vasily-kirichenko) [inline vsintegration/src/FSharp.Editor/Common/CommonHelpers.fs]: I'm 100% with you on this. It must be covered by tests, then rewritten in a more functional and safe manner. The number of things this algorithm operates is not as huge to make the code so imperative.

**Pull request merged**